### ISSUE TITLE
Compatibility with Draconic evolution mod

### ISSUE THREAD

**Issue by Libright1558** [pWn3d1337/Techguns2#106]
Techguns' weapons can bypass draconic evolution armor's shield which is supposed to block damage.

Is that intentional or just a Bug ?

Minecraft version : 1.12.2
Draconic evolution mod version : 1.12.2-2.3.20.333-universal
Techguns mod version : 2.0.1.2_1

**Comment by pWn3d1337** [comment]
If mod A changes things and mod B changes thing the results are unpredictable. There is no specific code that somehow tries to bypass it. I would call it a "compatibility conflict". Very likely the damage bypass that techguns uses for the different damage types is incompatible with whatever draconic evolution does.

**Comment by TerdyTheTerd** [comment]
Isn't the issue caused from [this](https://github.com/pWn3d1337/Techguns2/blob/master/src/main/java/techguns/damagesystem/DamageSystem.java#L425) event listener? Specifically where the EntityLivingBase has its health set, since DE listens for the [OnPlayerAttacked](https://github.com/brandon3055/Draconic-Evolution/blob/master/src/main/java/com/brandon3055/draconicevolution/handlers/CustomArmorHandler.java#L44) could we instead just call that with the calculated dmg instead of setting the hp if elb instanceof EntityPlayer so like `elb.attackEntitiyFrom(damageSrc.causePlayerDamage(damageSrc.getTrueSource()), damageAmount)`
Without actually going through your code and seeing just what you are overriding, I would assume a few additional checks to make sure the damage was from a player and not a turret (fake player?). I am also unsure about any additional checks or code needed, as I don't really have experience with forge modding, but I will set up a workspace and test this out, as I too would like to see some better compatability with DE, especially since DE listens for a forgevent which _should_ be fired when a player gets attacked

**Comment by TerdyTheTerd** [comment]
I was looking into this, the issue is that indeed mod A changes things mod B expects, except that in this case the bad actor, mod A, is TechGuns. In [TGEventHandler](https://github.com/pWn3d1337/Techguns2/blob/master/src/main/java/techguns/events/TGEventHandler.java#L344) you listen and cancel the OnLivingAttack event, with priority set to HIGH, an instead call your custom TGDamageSystem class. DE never gets to get activate it's shield, since its [CustomArmorHandler](https://github.com/brandon3055/Draconic-Evolution/blob/master/src/main/java/com/brandon3055/draconicevolution/handlers/CustomArmorHandler.java#L44) listens for the OnPlayerAttacked event (which your mod cancels, and never refires as it should). I have been testing and it seems that within DamageSystem I can simply re-fire the event, once all the custom damagesystem is done evaluating its final damage amount, obviously checking for target and attacker being a player ect. Should be submitting a PR once I've thoroughly tested it.

**Comment by TerdyTheTerd** [comment]
Well I have a solution that works beautifully, however since in my case I am removing turrets from the mod (my pack has a better turret mod already) I don't need to handle cases where the getTrueSource is not a player (turrets) so my changes won't work with the mod as a whole. I might look into a complete solution for others to benefit from, but it's not a priority of mine.

**Comment by AndreaS992** [comment]
I would like to have a fix or workaround for this issues as well.. It's pretty annoying that people can kill me even if I have a full draconic armor.

**Comment by TerdyTheTerd** [comment]
Seems I should go ahead and complete my patch to be more complete, as there are some other issues/suggestions I would like to implement anyways, might as well publish them for others to enjoy instead of hogging updates to my pack.

**Comment by Kodiax1** [comment]
I've got a modpack I've put together for some friends of mine, based off of FTB Revelations. I added in Techguns, and I'm wondering how precisely to get around the issue. I'd really appreciate it if you could share out your patch for it, even if the turrets get messed up along the way. It'd be incredibly appreciated by my friends and I.

**Comment by TerdyTheTerd** [comment]
Ahh ya Ill have to get around to getting those changes posted, I had begun to rework the damage system, so I have to go back and just pull the changes for the armor tweaks

**Comment by TerdyTheTerd** [comment]
My conundrum is this: I could PR the changes, but since they don't work with all aspects of the mod (turrets) it likely wont get accepted, plus PR's are a pain cuz of style nitpicking and slow acceptance. At the same time, I don't want to post my changes to my fork because players will go and download that version and will likely report issues in the wrong repo, potentially causing that repo's owner chasing down bugs that aren't even in their code.
It seems the best way would for me to release a compatibility mod for the few who really want it? It should be noted tho that I have not tested my changed with the shielding mechanics from this mods armor, and turrets also do not work because the server I am running has both of these disabled (other mods do them better).

**Comment by Kodiax1** [comment]
I was taking a (very amateurish) crack at it myself, just because I want
things to work the way they're supposed to. The server I play on has no use
for techgun's turrets, so if it breaks those, that's no loss to us. If
there's a concern the changes will break techgun's armor, I think we're
also alright moving away from that too, in favor of things like DE armor or
Avaritia armor. Would there be some way to get the changes to me, so our
server can do some testing for you and see what breaks and what doesn't?

**Comment by MasterZero** [comment]
I've build current version of techguns  #(4e26523cb0556ad56fefbee83c4df3fe94a426be) with your changes @TerdyTheTerd. I really appreciate what you made, thank you. People who wants that fix can find it here *removed*

**Comment by MasterZero** [comment]
I have better solution for damage system so it will handle damage as it should in both mods. Please, visit my [repo](https://github.com/MasterZero/Techguns2/releases/tag/2.0.2.0_pre3.2_de_compat) who still searching for solution.

**Comment by dvdcorsair** [comment]
@MasterZero could You add the patched version of mod on curseforge, please?

**Comment by MasterZero** [comment]
@dvdcorsair no, coz mod author do not give any permissions and "don't give a f*ck about your dildonic evolution"
